# Урок 4

# Системы линейных уравнений. Часть 1

__1.__ Решите систему уравнений методом Гаусса:

$$\begin{cases}
x_{1}+x_{2}-x_{3}-2x_{4}=0, \\
2x_{1}+x_{2}-x_{3}+x_{4}=-2, \\
x_{1}+x_{2}-3x_{3}+x_{4}=4.
\end{cases}$$

__Решение__

Запишем расширенную матрицу системы:

$$\begin{pmatrix}
\left.\begin{matrix}
1 & 1 & -1 & -2 \\ 
2 & 1 & -1 & 1 \\ 
1 & 1 & -3 & 1
\end{matrix}\right|
\begin{matrix}
0\\ 
-2\\
4
\end{matrix}
\end{pmatrix}.$$

В системе 4 переменных и 3 уравнения. _Возможно бесконечное множество решений._ 

Путём элементарных преобразований постараемся привести матрицу к ступенчатому виду.



Из второй строки вычтем первую, умноженную на 2.

$$\begin{pmatrix}
\left.\begin{matrix}
1 & 1 & -1 & -2 \\ 
0 & -1 & 1 & 5 \\ 
1 & 1 & -3 & 1
\end{matrix}\right|
\begin{matrix}
0\\ 
-2\\
4
\end{matrix}
\end{pmatrix}.$$

Из третьей строки вычтем первую.


$$\begin{pmatrix}
\left.\begin{matrix}
1 & 1 & -1 & -2 \\ 
0 & -1 & 1 & 5 \\ 
0 & 0 & -2 & 3
\end{matrix}\right|
\begin{matrix}
0\\ 
-2\\
4
\end{matrix}
\end{pmatrix}.$$

Оставшаяся матрица соответствует системе:

$$\begin{cases}
x_{1}+x_{2}-x_{3} -2x_{4}=0, \\
~~-x_{2}+x_{3} +5x_{4}=-2, \\
~~~~-2x_{3}+3x_{4}=4.
\end{cases}$$

И имеет бесконечное множество решений. Запишем в виде общего решения. 

$$x_{4}=c,$$
$$x_{3}=\frac{3c - 4}{2},$$
$$ -x_{2}+\frac{3c - 4}{2} +5c=-2 \; \Leftrightarrow \; x_{2}=\frac{13c}{2},$$
$$x_{1}+\frac{13c}{2} -\frac{3c - 4}{2} -2c=0 \; \Leftrightarrow \; x_{1}=-3c - 2.$$

Параметр $c$ может принимать бесконечное количество значений, при которых уравнения в системе будут обращаться в тождества.

Подставив общее решение в исходную систему, также получим тождества:

$$\begin{cases}
x_{1}+x_{2}-x_{3}-2x_{4}=0, \\
2x_{1}+x_{2}-x_{3}+x_{4}=-2, \\
x_{1}+x_{2}-3x_{3}+x_{4}=4.
\end{cases}$$

$$\begin{cases}
-3c - 2 +\frac{13c}{2} - \frac{3c - 4}{2} -2c =0, \\
2(-3c - 2)+\frac{13c}{2}-\frac{3c - 4}{2}+c=-2, \\
-3c - 2+\frac{13c}{2}-3\frac{3c - 4}{2}+c=4.
\end{cases}$$

__Ответ__:

$$x_{1}=-3c - 2,$$
$$x_{2}=\frac{13c}{2},$$
$$x_{3}=\frac{3c - 4}{2},$$
$$x_{4}=c.$$



__2.__ Проверьте на совместность и выясните, сколько решений будет иметь система линейных уравнений:

   а) $\begin{cases}
3x_{1}-x_{2}+x_{3}=4, \\
2x_{1}-5x_{2}-3x_{3}=-17, \\
x_{1}+x_{2}-x_{3}=0;
\end{cases}$
    
   б) $\begin{cases}
2x_{1}-4x_{2}+6x_{3}=1, \\
x_{1}-2x_{2}+3x_{3}=-2, \\
3x_{1}-6x_{2}+9x_{3}=5;
\end{cases}$
    
   в) $\begin{cases}
x_{1}+2x_{2}+5x_{3}=4, \\
3x_{1}+x_{2}-8x_{3}=-2. 
\end{cases}$


__Решение__

   а) $\begin{cases}
3x_{1}-x_{2}+x_{3}=4, \\
2x_{1}-5x_{2}-3x_{3}=-17, \\
x_{1}+x_{2}-x_{3}=0;
\end{cases}$

$$\begin{pmatrix}
\left.\begin{matrix}
3 & -1 & 1 \\ 
2 & -5 & -3\\ 
1 & 1 & -1
\end{matrix}\right|
\begin{matrix}
4\\ 
-17\\
0
\end{matrix}
\end{pmatrix}.$$

__Когда число уравнений равно числу неизвестных__ и матрица коэффициентов не вырождена, если векторы столбцов коэффициентов $a_{1},a_{2}...,a_{n}$ линейно независимы, то такая система совместна, и её решение единственно при любом векторе свободных членов $b$._ 

В нашем случае, число уравнений равно числу неизвестных, вектора столбцов линейно независимы, соответственно такая __система совместна и имеет одно решение__.

Проверим методом Гаусса:

$$\begin{pmatrix}
\left.\begin{matrix}
3 & -1 & 1 \\ 
2 & -5 & -3\\ 
1 & 1 & -1
\end{matrix}\right|
\begin{matrix}
4\\ 
-17\\
0
\end{matrix}
\end{pmatrix}.$$

Из второй строки вычтем первую, умноженную на 2/3

$$\begin{pmatrix}
\left.\begin{matrix}
3 & -1 & 1 \\ 
0 & -13/3 & -11/3\\ 
1 & 1 & -1
\end{matrix}\right|
\begin{matrix}
4\\ 
-59/3\\
0
\end{matrix}
\end{pmatrix}.$$

Из третьей строки вычтем первую, умноженную на 1/3

$$\begin{pmatrix}
\left.\begin{matrix}
3 & -1 & 1 \\ 
0 & -13/3 & -11/3\\ 
0 & 4/3 & -4/3
\end{matrix}\right|
\begin{matrix}
4\\ 
-59/3\\
-4/3
\end{matrix}
\end{pmatrix}.$$

К третьей строке прибавим вторую, умноженную на 4/13

$$\begin{pmatrix}
\left.\begin{matrix}
3 & -1 & 1 \\ 
0 & -13/3 & -11/3\\ 
0 & 0 & (-4/3 - 44/39)
\end{matrix}\right|
\begin{matrix}
4\\ 
-59/3\\
(-4/3 - 236/39)
\end{matrix}
\end{pmatrix}.$$

__Действительно, получилось, что система совместна и имеет одно решение.__

    
   б) $\begin{cases}
2x_{1}-4x_{2}+6x_{3}=1, \\
x_{1}-2x_{2}+3x_{3}=-2, \\
3x_{1}-6x_{2}+9x_{3}=5;
\end{cases}$

$$\begin{pmatrix}
\left.\begin{matrix}
2 & -4 & 6 \\ 
1 & -2 & 3\\ 
3 & -6 & 9
\end{matrix}\right|
\begin{matrix}
1\\ 
-2\\
5
\end{matrix}
\end{pmatrix}.$$

В данном случае мы видим, что в матрице коэффициентов, первая и вторая строки линейно зависимы, если одну из них их убрать, то ранг матрицы равен 2. В случае расширенной матрицы, строки линейно независимы, то есть ранг матрицы = 3. 
Третий столбец равен первому, умноженному на 3 (то есть лежат на одной прямой). Столбец свободных членов нельзя выразить через столбцы коэффициентов, то есть вектор свободных членов не лежит в плоскости, образуемой векторами коэффициентов.
Соответственно, __система несовместна__.


   в) $\begin{cases}
x_{1}+2x_{2}+5x_{3}=4, \\
3x_{1}+x_{2}-8x_{3}=-2. 
\end{cases}$

$$\begin{pmatrix}
\left.\begin{matrix}
1 & 2 & 5 \\ 
3 & 1 & -8
\end{matrix}\right|
\begin{matrix}
4\\ 
-2
\end{matrix}
\end{pmatrix}.$$

Число уравнений меньше числа неизвестных, система недоопределена.
Вектора коэфициентов линейно независимы. 
Ранг и матрицы коэффициентов, и расширенной матрицы равен 2, что меньше числа переменных, которых 3.
Соответственно, __система имеет бесконечное количество решений__.

In [175]:
#Проверим с помощью Python

import numpy as np

In [176]:
def gauss_method(A, b, n):
    rnk_A = np.linalg.matrix_rank(A)
    rnk_Ab = np.linalg.matrix_rank(np.hstack((A, b)))
    if  rnk_A == rnk_Ab == n:
        return 'Cистема имеет единственное решение'
    elif (rnk_A == rnk_Ab) and (rnk_A < n):
        return 'Cистема имеет бесконечное количество решений'
    elif rnk_A < rnk_Ab:
        return 'Cистема несовместна'

In [177]:
A = np.array([[3, -1 , 1], [2, -5, -3], [1, 1, -1]])
b = np.array([4, -17, 0]).reshape(-1, 1)
gauss_method(A, b, 3)

'Cистема имеет единственное решение'

In [178]:
A = np.array([[2, -4, 6], [1, -2, 3], [3, -6, 9]])
b = np.array([1, -2, 5]).reshape(-1, 1)
gauss_method(A, b, 3)

'Cистема несовместна'

In [179]:
A = np.array([[1, 2, 5], [3, 1, -8]])
b = np.array([4, -2]).reshape(-1, 1)
gauss_method(A, b, 3)

'Cистема имеет бесконечное количество решений'

__3.__ Проверьте на совместность и выясните, сколько решений будет иметь система линейных уравнений, заданная расширенной матрицей:

$$\tilde{A}=\begin{pmatrix}
\left.\begin{matrix}
1 & 3 & -2 & 4\\ 
0 & 5 & 0 & 1\\ 
0 & 0 & 3 & 0\\ 
0 & 0 & 0 & 2
\end{matrix}\right|
\begin{matrix}
3\\ 
2\\
4\\
1
\end{matrix}
\end{pmatrix}.$$

__Решение__

In [180]:
A = np.array([[1, 3, -2, 4], [0, 5, 0, 1], [0, 0, 3, 0], [0, 0, 0, 2]])
b = np.array([3, 2, 4, 1]).reshape(-1, 1)
gauss_method(A, b, 4)

'Cистема имеет единственное решение'

__Ответ__: система совместна и имеет единственное решение.

__4.__ Дана система линейных уравнений, заданная расширенной матрицей:

$$\tilde{A}=\begin{pmatrix}
\left.\begin{matrix}
1 & 2 & 3\\ 
4 & 5 & 6\\ 
7 & 8 & 9
\end{matrix}\right|
\begin{matrix}
a\\ 
b\\
c
\end{matrix}
\end{pmatrix}.$$

Найдите соотношение между параметрами $a$, $b$ и $c$, при которых система считается несовместной.

__Решение__


По теореме Кронекера — Капелли:

Необходимым и достаточным условием совместности системы из $m$ уравнений с $n$ неизвестными считается равенство между собой рангов матрицы коэффициентов $A$ и расширенной матрицы $\tilde A$:

$$rank A=rank \tilde A.$$

Причём:<br>
* если $rankA=rank \tilde A=n$, где $n$ — число неизвестных, то система определена, то есть имеет единственное решение;<br>
* если $rankA=rank \tilde A<n$, то система имеет бесконечное количество решений;<br>
* если $rankA<rank \tilde A$, то система несовместна.

__То есть получается, если ранг матрицы коэффициентов меньше ранга расширенной матрицы, то система несовместна.__ 

In [181]:
#Найдем ранг матрицы коэффициентов

A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])

rnk_A = np.linalg.matrix_rank(A)
rnk_A

2

Получается, нам надо добиться того, чтобы ранг расширенной матрицы стал больше ранга матрицы, коэффеициентов. 

$$\tilde{A}=\begin{pmatrix}
\left.\begin{matrix}
1 & 2 & 3\\ 
4 & 5 & 6\\ 
7 & 8 & 9
\end{matrix}\right|
\begin{matrix}
a\\ 
b\\
c
\end{matrix}
\end{pmatrix}.$$

Преобразуем матрицу методом Гаусса:

От 2-й строки отнимем 1-ю, умноженную на 4

$$\tilde{A}=\begin{pmatrix}
\left.\begin{matrix}
1 & 2 & 3\\ 
0 & -3 & -6\\ 
7 & 8 & 9
\end{matrix}\right|
\begin{matrix}
a\\ 
b-4a\\
c
\end{matrix}
\end{pmatrix}.$$



От 3-й строки отнимем 1-ю, умноженную на 7

$$\tilde{A}=\begin{pmatrix}
\left.\begin{matrix}
1 & 2 & 3\\ 
0 & -3 & -6\\ 
0 & -6 & -12
\end{matrix}\right|
\begin{matrix}
a\\ 
b-4a\\
c-7a
\end{matrix}
\end{pmatrix}.$$


Получается, третья строка матрицы коэффициентов равна второй, умноженнной на 2. Из-за этого ранг матрицы = 2.


То есть надо с помощью соотношения коэфициентов, добиться того, чтобы 3я строка не была пропорциональна второй строке. 

Нужное соотношение является результатом решения неравенства:

$$ c - 7a \neq 2(b - 4a) \Leftrightarrow \; c \neq 2b - a \Leftrightarrow \; a \neq  2b - c \Leftrightarrow \; b \neq  \frac{c+a}{2} $$

Подставим пример, соответствующий данному соотношению в исходную матрицу и найдем ее ранг.

Подствляем a=1, b=2, $с \neq 2b - a \neq 3$. Подставим c=4
$$\tilde{A}=\begin{pmatrix}
\left.\begin{matrix}
1 & 2 & 3\\ 
4 & 5 & 6\\ 
7 & 8 & 9
\end{matrix}\right|
\begin{matrix}
1\\ 
2\\
4
\end{matrix}
\end{pmatrix}.$$

In [182]:

A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9], [1, 2, 4],])

rnk_A = np.linalg.matrix_rank(A)
rnk_A

3

__Ранг расширенной матрицы больше, ранга матрицы коэффициентов, то есть система несовместна.__

In [183]:
# если поставить с=3, то ранг будет 2 и система будет иметь бесконечное количество решений, так как ранги равны, но меньше
# числа переменных.
A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9], [1, 2, 3],])

rnk_A = np.linalg.matrix_rank(A)
rnk_A

2

__Ответ__:
Соотношение следующее:

$$ c \neq 2b - a \Leftrightarrow \; a \neq  2b - c \Leftrightarrow \; b \neq  \frac{c+a}{2} $$

# Системы линейных уравнений. Часть 2

__1.__ Решите систему уравнений методом Крамера:

   а) $\begin{cases}
x_{1}-2x_{2}=1 \\
3x_{1}-4x_{2}=7
\end{cases}$
    
   б) $\begin{cases}
2x_{1}-x_{2}+5x_{3}=10 \\
x_{1}+x_{2}-3x_{3}=-2 \\
2x_{1}+4x_{2}+x_{3}=1
\end{cases}$

__Решение__

   а) $\begin{cases}
x_{1}-2x_{2}=1 \\
3x_{1}-4x_{2}=7
\end{cases}$

Для системы с $n$ уравнениями и $n$ неизвестными, если система уравнений не вырождена ($detA\neq 0$), то система определена, то есть имеет единственное решение, и это решение может быть найдено по формуле Крамера:

$$x_{i}=\frac{detA_{i}}{detA},$$

где $A_{i}$ — это матрица, получаемая заменой $i$-го столбца на вектор-столбец свободных членов $b$.

$$\tilde{A}=\begin{pmatrix}
\left.\begin{matrix}
1 & -2\\ 
3 & -4\\ 
\end{matrix}\right|
\begin{matrix}
1\\ 
7
\end{matrix}
\end{pmatrix}.$$

$$detA = (1 * (-4)) - 3 * (-2) = -4 + 6 = 2$$

In [184]:
A = np.array([[1,-2], [3, -4]])
np.linalg.det(A)

2.0000000000000004

$$detA_1 = (1 * (-4)) - (-2) * 7 = -4 + 14 = 10$$

In [185]:
A_1 = np.array([[1, -2], [7, -4]])
np.linalg.det(A_1)

9.999999999999998

$$detA_2 = (1 * 7) - 1 * 3 = 7 - 3 = 4$$

In [186]:
A_2 = np.array([[1, 1], [3, 7]])
np.linalg.det(A_2)

3.999999999999999

$$x_{1}=\frac{10}{2} = 5$$


In [187]:
np.linalg.det(A_1) / np.linalg.det(A)

4.999999999999998

$$x_{2}=\frac{4}{2} = 2$$

In [188]:
np.linalg.det(A_2) / np.linalg.det(A)

1.9999999999999991

__Ответ:__

x1 = 5, x2 = 2

   б) $\begin{cases}
2x_{1}-x_{2}+5x_{3}=10 \\
x_{1}+x_{2}-3x_{3}=-2 \\
2x_{1}+4x_{2}+x_{3}=1
\end{cases}$

$$\tilde{A}=\begin{pmatrix}
\left.\begin{matrix}
2 & -1 & 5\\ 
1 & 1 & -3\\ 
2 & 4 & 1\\
\end{matrix}\right|
\begin{matrix}
10\\ 
-2\\
1
\end{matrix}
\end{pmatrix}.$$

$$x_{i}=\frac{detA_{i}}{detA},$$

$$detA = 2 (1 + 12) + 1 (1 + 6) + 5 (4 - 2) = 42$$

In [189]:
A = np.array([[2,-1, 5], [1, 1, -3], [2, 4, 1]])
np.linalg.det(A)

42.99999999999998

In [190]:
A_1 = np.array([[10,-1, 5], [-2, 1, -3], [1, 4, 1]])
np.linalg.det(A_1)

86.00000000000004

In [191]:
A_2 = np.array([[2, 10, 5], [1, -2, -3], [2, 1, 1]])
np.linalg.det(A_2)

-43.000000000000014

In [192]:
A_3 = np.array([[2,-1, 10], [1, 1, -2], [2, 4, 1]])
np.linalg.det(A_3)

43.0

In [193]:
x_1  = np.linalg.det(A_1)/ np.linalg.det(A)
x_1

2.0000000000000018

In [194]:
x_2  = np.linalg.det(A_2)/ np.linalg.det(A)
x_2

-1.0000000000000009

In [195]:
x_3  = np.linalg.det(A_3)/ np.linalg.det(A)
x_3

1.0000000000000004

Подставим полученные значения в исходное уравнение:

$ 2 * 2 -(-1)+5 * 1 =10$ 

$ 4 + 1 +5 * 1 =10$ 

Равенство выполняется.

__Ответ:__

x1 = 2, x2 = -1, x3 = 1


__2*.__ Найдите $L$-матрицу $LU$-разложения для матрицы коэффициентов:

   а)$$\begin{pmatrix}
1 & 2 & 4 \\ 
2 & 9 & 12 \\ 
3 & 26 & 30
\end{pmatrix}$$
    
   б)$$\begin{pmatrix}
1 & 1 & 2 & 4\\ 
2 & 5 & 8 & 9\\ 
3 & 18 & 29 & 18\\
4 & 22 & 53 & 33
\end{pmatrix}$$

__Решение__

   а)$$\begin{pmatrix}
1 & 2 & 4 \\ 
2 & 9 & 12 \\ 
3 & 26 & 30
\end{pmatrix}$$

__Теорема__

Пусть дана квадратная матрица $A$ порядка $n$, и $A_{k}$ — главный минор этой матрицы, составленный из первых $k$ строк и столбцов. Если $det(A_{k})\neq0$ для $k=1,2...,n-1$, тогда существует единственная нижняя треугольная матрица $L=(l_{ij})$, где $l_{11}=l_{22}=...=l_{nn}=1$, и единственная верхняя треугольная матрица $U=(u_{ij})$, что $LU=A$. Более того, $det(A)=u_{11}\cdot u_{22}\cdot...\cdot u_{nn}$.

Представление матрицы системы уравнений $A$ в виде произведения $LU$ считается основной идеей гауссовских схем исключения, так как система $Ax=b$ может быть записана как:

$$LUx=b$$

и сводится к двум системам с треугольными матрицами.

$L$ - нижняя треугольная матрица, $U$ - верхнетреугольная матрица, получаемая путем классического метода Гаусса.

Верхняя треугольная матрица $U$ получается путём прямого хода метода Гаусса.

Повторяя алгоритм из предыдущего урока, если $a_{11}\neq0$, на первом шаге умножаем первое уравнение на 

$$m_{i1}=\frac{a_{i1}}{a_{11}}$$

и вычитаем его из $i$-го ($i=2,3,4$).

В итоге получаем эквивалентную систему

$$\begin{pmatrix}
a_{11} & a_{12} & a_{13} & a_{14}\\ 
0 & a_{22}^{(1)} & a_{23}^{(1)} & a_{24}^{(1)}\\ 
0 & a_{32}^{(1)} & a_{33}^{(1)} & a_{34}^{(1)}\\ 
0 & a_{42}^{(1)} & a_{43}^{(1)} & a_{44}^{(1)}\\ 
\end{pmatrix}
\begin{pmatrix}
x_{1}\\ 
x_{2}\\ 
x_{3}\\ 
x_{4}\\ 
\end{pmatrix}=
\begin{pmatrix}
b_{1}\\ 
b_{2}^{1}\\ 
b_{3}^{1}\\ 
b_{4}^{1}\\ 
\end{pmatrix},$$

где:

$$a_{ij}^{(1)}=a_{ij}-a_{1j}\frac{a_{i1}}{a_{11}}~~~~(i,j\geq2),$$
$$b_{i}^{(1)}=b_{i}-b_{1}\frac{a_{i1}}{a_{11}}.$$

$L$ находится следующим образом:

$$L=\begin{pmatrix}
1 & 0 & 0 & 0\\ 
m_{21} & 1 & 0 & 0\\ 
m_{31} & m_{32} & 1 & 0\\ 
m_{41} & m_{42} & m_{43} & 1\\ 
\end{pmatrix}.$$

Проделаем эти операции. Найдем матрицу $U$ и соответствующие коэффициенты.

$$\begin{pmatrix}
1 & 2 & 4 \\ 
2 & 9 & 12 \\ 
3 & 26 & 30
\end{pmatrix}$$

Отнимем из второй строки первую умноженную на 2, то есть $m_{21} = 2$

$$\begin{pmatrix}
1 & 2 & 4 \\ 
0 & 5 & 4 \\ 
3 & 26 & 30
\end{pmatrix}$$

Отнимем из третьей строки первую умноженную на 3, то есть $m_{31} = 3$

$$\begin{pmatrix}
1 & 2 & 4 \\ 
0 & 5 & 4 \\ 
0 & 20 & 18
\end{pmatrix}$$

Отнимем из третьей строки вторую умноженную на 4, то есть $m_{32} = 4$

$$\begin{pmatrix}
1 & 2 & 4 \\ 
0 & 5 & 4 \\ 
0 & 0 & 2
\end{pmatrix}$$

Получили матрицу $U$:

$$\begin{pmatrix}
1 & 2 & 4 \\ 
0 & 5 & 4 \\ 
0 & 0 & 2
\end{pmatrix}$$

Найдем матрицу $L$:

$$\begin{pmatrix}
1 & 0 & 0 \\ 
2 & 1 & 0 \\ 
3 & 4 & 1
\end{pmatrix}$$

$$A=LU.$$

Проверим на практике - все верно. 

In [196]:
L = np.array([[1, 0, 0], [2, 1, 0], [3, 4, 1]])
U = np.array([[1, 2, 4], [0, 5, 4], [0, 0, 2]])
np.dot(L, U)

array([[ 1,  2,  4],
       [ 2,  9, 12],
       [ 3, 26, 30]])

   б)$$\begin{pmatrix}
1 & 1 & 2 & 4\\ 
2 & 5 & 8 & 9\\ 
3 & 18 & 29 & 18\\
4 & 22 & 53 & 33
\end{pmatrix}$$

Найдем матрицу $U$ и соответствующие коэффициенты.

$$\begin{pmatrix}
1 & 1 & 2 & 4\\ 
2 & 5 & 8 & 9\\ 
3 & 18 & 29 & 18\\
4 & 22 & 53 & 33
\end{pmatrix}$$

Для экономии времени воспользуемся Python

In [197]:
A = np.array([[1, 1, 2, 4], [2, 5, 8, 9], [3, 18, 29, 18], [4, 22, 53, 33]])

In [198]:
# Отнимем из второй строки первую умноженную на 2, то есть m_21 = 2
A[1,] = A[1,] - A[0,]*2
A

array([[ 1,  1,  2,  4],
       [ 0,  3,  4,  1],
       [ 3, 18, 29, 18],
       [ 4, 22, 53, 33]])

In [199]:
# Отнимем из третьей строки первую, умноженную на 3, то есть m_31 = 3
A[2,] = A[2,] - A[0,]*3
A

array([[ 1,  1,  2,  4],
       [ 0,  3,  4,  1],
       [ 0, 15, 23,  6],
       [ 4, 22, 53, 33]])

In [200]:
# Отнимем из четвертой строки первую, умноженную на 4, то есть m_41 = 4
A[3,] = A[3,] - A[0,]*4
A

array([[ 1,  1,  2,  4],
       [ 0,  3,  4,  1],
       [ 0, 15, 23,  6],
       [ 0, 18, 45, 17]])

In [201]:
# Отнимем из третьей строки вторую, умноженную на 5, то есть m_32 = 5
A[2,] = A[2,] - A[1,]*5
A

array([[ 1,  1,  2,  4],
       [ 0,  3,  4,  1],
       [ 0,  0,  3,  1],
       [ 0, 18, 45, 17]])

In [202]:
# Отнимем из четвертой строки вторую, умноженную на 6, то есть m_42 = 6
A[3,] = A[3,] - A[1,]*6
A

array([[ 1,  1,  2,  4],
       [ 0,  3,  4,  1],
       [ 0,  0,  3,  1],
       [ 0,  0, 21, 11]])

In [203]:
# Отнимем из четвертой строки третью, умноженную на 7, то есть m_43 = 7
A[3,] = A[3,] - A[2,]*7
A

array([[1, 1, 2, 4],
       [0, 3, 4, 1],
       [0, 0, 3, 1],
       [0, 0, 0, 4]])

In [204]:
#U матрица готова
U = A
U

array([[1, 1, 2, 4],
       [0, 3, 4, 1],
       [0, 0, 3, 1],
       [0, 0, 0, 4]])

In [205]:
#теперь построим L матрицу с помощью коэффициентов

L = np.array([[1, 0, 0, 0], [2, 1, 0, 0], [3, 5, 1, 0], [4, 6, 7, 1]])
L 

array([[1, 0, 0, 0],
       [2, 1, 0, 0],
       [3, 5, 1, 0],
       [4, 6, 7, 1]])

In [206]:
A = np.dot(L, U)
A

array([[ 1,  1,  2,  4],
       [ 2,  5,  8,  9],
       [ 3, 18, 29, 18],
       [ 4, 22, 53, 33]])

In [207]:
#Итог, L матрица получилась следующей:
L

array([[1, 0, 0, 0],
       [2, 1, 0, 0],
       [3, 5, 1, 0],
       [4, 6, 7, 1]])

__3*.__ Решите систему линейных уравнений методом $LU$-разложения:

$$\begin{cases}
2x_{1}+x_{2}+3x_{3}=1 \\
11x_{1}+7x_{2}+5x_{3}=-6 \\
9x_{1}+8x_{2}+4x_{3}=-5
\end{cases}$$

__Решение__

Матрица коэффициентов $A$ примет вид:

$$\begin{pmatrix}
2 & 1 & 3 \\ 
11 & 7 & 5 \\ 
9 & 8 & 4
\end{pmatrix}.$$

In [208]:
#Найдем U-матрицу путем прямого хода Гаусса.
A = np.array([[2, 1, 3], [11, 7, 5], [9, 8, 4]], dtype=np.float64)
A

array([[ 2.,  1.,  3.],
       [11.,  7.,  5.],
       [ 9.,  8.,  4.]])

In [209]:
#разделим первую строку на 2 и домножим на 11 и отнимем от второй m21 = 11/2
A[1,] = A[1,] - (A[0,]*11/2)
A

array([[  2. ,   1. ,   3. ],
       [  0. ,   1.5, -11.5],
       [  9. ,   8. ,   4. ]])

In [210]:
#разделим первую строку на 2 и домножим на 9 и отнимем от третьей m31 = 9/2
A[2,] = A[2,] - (A[0,]*9/2)
A

array([[  2. ,   1. ,   3. ],
       [  0. ,   1.5, -11.5],
       [  0. ,   3.5,  -9.5]])

In [211]:
#разделим вторую строку на 1,5 и домножим на 3,5 и отнимем от третьей m32 = 3.5/1.5
A[2,] = A[2,] - (A[1,]*3.5/1.5)
A

array([[  2.        ,   1.        ,   3.        ],
       [  0.        ,   1.5       , -11.5       ],
       [  0.        ,   0.        ,  17.33333333]])

In [212]:
# U матрица готова
U = A 
U

array([[  2.        ,   1.        ,   3.        ],
       [  0.        ,   1.5       , -11.5       ],
       [  0.        ,   0.        ,  17.33333333]])

In [213]:
#построим L матрицу
L = np.array([[1, 0, 0], [11/2, 1, 0], [9/2, 3.5/1.5, 1]], dtype=np.float64)
L

array([[1.        , 0.        , 0.        ],
       [5.5       , 1.        , 0.        ],
       [4.5       , 2.33333333, 1.        ]])

In [214]:
#проверяем
np.dot(L, U)

array([[ 2.,  1.,  3.],
       [11.,  7.,  5.],
       [ 9.,  8.,  4.]])

Теперь решим систему:

$$Ly=b:$$

$$\begin{cases}
y_{1}=1, \\
5.5y_{1}+y_{2}=-6, \\
4.5y_{1}+ 2.33y_{2}+y_{3}=-5.
\end{cases}$$

$$y_{1}=1,$$
$$y_{2}=-11.5,$$
$$y_{3}=17.295.$$

In [215]:
U

array([[  2.        ,   1.        ,   3.        ],
       [  0.        ,   1.5       , -11.5       ],
       [  0.        ,   0.        ,  17.33333333]])

Затем систему:

$$Ux=y:$$

$$\begin{cases}
2x_{1}+x_{2}+ 3x_{3}=1, \\
1.5x_{2} - 11.5x_{3}=-11.5, \\
17.33x_{3}=17.295.
\end{cases}$$

$$x_{3} \approx 0,998 \approx 1$$
$$x_{2}\approx 0,$$
$$x_{1}\approx -1.$$

Подставим решение в исходную систему:

$$\begin{cases}
2(-1)+ 0 +3 * 1 =1 \\
11(-1)+7 * 0 +5 * 1 =-6 \\
9(-1)+8 * 0 +4 * 1 = -5
\end{cases}$$

Равенство выполняется. 

__Ответ__
$$x_{3} \approx 0,998 \approx 1$$
$$x_{2}\approx 0,$$
$$x_{1}\approx -1.$$


__4*.__ Решите систему линейных уравнений методом Холецкого:

$$\begin{cases}
81x_{1}-45x_{2}+45x_{3}=531 \\
-45x_{1}+50x_{2}-15x_{3}=-460 \\
45x_{1}-15x_{2}+38x_{3}=193
\end{cases}$$

__Решение__

$$\begin{pmatrix}
81 & -45 & 45 \\ 
-45 & 50 & -15 \\ 
45 & -15 & 38
\end{pmatrix}.$$

Произведём разложение на $LL^{T}$:

$$l_{11}=\sqrt{a_{11}}=\sqrt{81} = 9,$$
$$l_{21}=\frac{a_{21}}{l_{11}}=\frac{-45}{9} = - 5,$$
$$l_{31}=\frac{a_{31}}{l_{11}}=\frac{45}{9}=5,$$

$$l_{22}=\sqrt{a_{22}-l_{21}^{2}}=\sqrt{50 - 25}= 5,$$
$$l_{32}=\frac{1}{l_{22}}\left ( a_{32}-l_{21}l_{31} \right)=\frac{1}{5}\left ( -15 - (-5)\cdot 5 \right )=2,$$

$$l_{33}=\sqrt{a_{33}-l_{31}^{2}-l_{32}^{2}}=\sqrt{38-25-4}=3.$$


Получили матрицу:

$$L = \begin{pmatrix}
9 & 0 & 0 \\ 
-5 & 5 & 0 \\ 
5 & 2 & 3
\end{pmatrix}, 
\; \; 
L^{T} = \begin{pmatrix}
9 & -5 & 5 \\ 
0 & 5 & 2 \\ 
0 & 0 & 3
\end{pmatrix}.$$

Решим систему $Ly=b:$

$$\begin{cases}
9y_{1}=531, \\
-5y_{1}+5y_{2}=-460, \\
5y_{1}+2y_{2}+3y_{3}=193.
\end{cases}$$

$$y_{1} = 59,$$

$$y_{2}=-33,$$

$$y_{3}=-12.$$

И решим систему $L^{T}x=y:$

$$\begin{cases}
9x_{1}-5x_{2}+5x_{3}=59, \\
5x_{2}+2x_{3}=-33, \\
3x_{3}=-12.
\end{cases}$$

$$x_{3}=-4,$$

$$x_{2}=-5,$$

$$x_{1}=6.$$

Осуществим проверку, подставив полученные значения в исходную систему:

$$\begin{cases}
81 \cdot 6 -45 \cdot (-5) +45 \cdot (-4)=531 \\
-45 \cdot 6 +50 \cdot (-5 ) - 15 \cdot (-4) =-460 \\
45  \cdot 6 -15 \cdot (-5) + 38  \cdot (-4)=193
\end{cases}$$

__Ответ__

$$x_{3}=-4,$$

$$x_{2}=-5,$$

$$x_{1}=6.$$

__5*.__ Напишите на Python программу с реализацией одного из изученных алгоритмов решения СЛАУ.

__Решение__

Напишем реализацию метода Гаусса.

In [216]:
"""Класс для решения СЛАУ, которые имеют единственное решение, методом Гаусса"""
class GaussMethod:
    def __init__(self, A, b):
        self.A = A
        self.b = b
        self.n = len(self.A)
        self.A_hat = np.hstack((self.A, self.b))

    """Проверка совместности"""
    def __consistency_check(self):
        rnk_A = np.linalg.matrix_rank(self.A)
        rnk_Ab = np.linalg.matrix_rank(np.hstack((self.A, self.b)))
        #'Cистема несовместна'
        if rnk_A < rnk_Ab:
            return -1
        #'Cистема имеет единственное решение'
        elif rnk_A == rnk_Ab == self.n:
            return 1
        #'Cистема имеет бесконечное количество решений'
        elif (rnk_A == rnk_Ab) and (rnk_A < self.n):
            return 2

    """Получение верхнетреугольной матрицы"""
    def _get_triangle_matrix(self):
        if self.__consistency_check() == -1 or self.__consistency_check() == 2:
            return -1
        else:
            M = self.A_hat.copy()
            for j in range(0, len(M)-1):
                for i in range(1, len(M)):
                    if i + j < len(M):
                        M[i+j, ] = M[i+j, ] - M[j, ] * (M[i+j, j] / M[j, j])
            return M

    """Нахождение неизвестных методом обратного хода Гаусса"""
    def gauss_reverse(self):
        M = self._get_triangle_matrix()
        if self.__consistency_check() == -1:
            print('Cистема несовместна')
            return -1
        if self.__consistency_check() == 2:
            print('Cистема имеет бесконечное количество решений')
            return -1
        elif self.__consistency_check() == 1:
            print('Cистема имеет единственное решение')
            M_short = M[:,:-1]
            b_short = M[:,-1].reshape(-1, 1)
            M_x = [0 for i in range(len(M_short))]
            for i in reversed(range(len(M_short))):
                x_str = M_short[i,]
                c = 0
                for idx, el in enumerate(x_str):
                    if idx != i:
                        c += el * M_x[idx]
                x = (b_short[i] - c) / x_str[i]
                M_x[i] = x[0]
            return M_x


# A = np.array([[2, 1, 3], [11, 7, 5], [9, 8, 4]], dtype=np.float64)
# b = np.array([1, -6, 5], dtype=np.float64).reshape(-1, 1)

# A = np.array([[2,-1, 5], [1, 1, -3], [2, 4, 1]], dtype=np.float64)
# b = np.array([10, -2, 1], dtype=np.float64).reshape(-1, 1)

# A = np.array([[1, 3, -2, 4], [0, 5, 0, 1], [0, 0, 3, 0], [0, 0, 0, 2]])
# b = np.array([3, 2, 4, 1]).reshape(-1, 1)

A = np.array([[1, 3, -2, 4, 8], [1, 5, 0, 1, 8], [7, 7, 3, 0, 8], [5, 5, 7, 2, 8], [1, 2, 3, 2, 8]])
b = np.array([3, 2, 4, 1, 7]).reshape(-1, 1)

matrix = GaussMethod(A, b)
print(matrix._get_triangle_matrix())
print(matrix.gauss_reverse())
        

[[  1   3  -2   4   8   3]
 [  0   2   2  -3   0  -1]
 [  0   0  31 -49 -48 -24]
 [  0   0   0   9   9   1]
 [  0   0   0   0   3   6]]
Cистема имеет единственное решение
[1.2401433691756267, -2.6702508960573472, -0.6630824372759857, -1.8888888888888888, 2.0]
